File: LeetCode/Easy/0141-linked-list-cycle/0141-linked-list-cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;ㅆ
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        // create a dictionary for value-index 
        HashMap<Integer, Integer> valueMap = new HashMap<Integer, Integer>();
        while(head != null) {
            if (valueMap.containsKey(head.hashCode())) {
                return true;
            }
            valueMap.put(head.hashCode(), 0);
            head = head.next;
        }
        return false;
    }
}

File: LeetCode/Easy/0141-linked-list-cycle/0141-linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        # hashamp solution / time: O(N), space: O(N)
        # visited = set()
        # current = head

        # while current:
        #     if current in visited:
        #         return True 
        #     visited.add(current)
        #     current = current.next
        # return False

        # two pointers (floy'd cycle detection algorithm )
        if not head or not head.next: 
            return False

        slow, fast = head, head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                return True 
        return False
